i18n-helpers/src/lib.rs

```diff
@@ -334,18 +336,129 @@ fn is_comment_skip_directive(html: &str) -> bool {
     re.is_match(html.trim())
 }
 
-/// Returns true if the events appear to be a codeblock without translatable text.
-fn is_nontranslatable_codeblock_group(events: &[(usize, Event)]) -> bool {
-    match events {
-        [(_, Event::Start(Tag::CodeBlock(_))), .., (_, Event::End(Tag::CodeBlock(_)))] => {
-            let (codeblock_text, _) = reconstruct_markdown(events, None);
-            // Heuristic to check whether the codeblock nether has a
-            // literal string nor a line comment.  We may actually
-            // want to use a lexer here to make this more robust.
-            !codeblock_text.contains('"') && !codeblock_text.contains("//")
+/// Returns true if the events appear to be a codeblock.
+fn is_codeblock_group(events: &[(usize, Event)]) -> bool {
+    matches!(
+        events,
+        [
+            (_, Event::Start(Tag::CodeBlock(_))),
+            ..,
+            (_, Event::End(Tag::CodeBlock(_)))
+        ]
+    )
+}
+
+/// Returns true if the scope should be translated.
+fn is_translate_scope(x: Scope) -> bool {
+    static SCOPE_STRING: OnceLock<Scope> = OnceLock::new();
+    static SCOPE_COMMENT: OnceLock<Scope> = OnceLock::new();
+
+    let scope_string = SCOPE_STRING.get_or_init(|| Scope::new("string").unwrap());
+    let scope_comment = SCOPE_COMMENT.get_or_init(|| Scope::new("comment").unwrap());
+    scope_string.is_prefix_of(x) || scope_comment.is_prefix_of(x)
+}
+
+/// Creates groups by parsing codeblock.
+fn parse_codeblock<'a>(events: &'a [(usize, Event)]) -> Vec<Group<'a>> {
+    // Language detection from language identifier of codeblock.
+    let ss = SyntaxSet::load_defaults_newlines();
+    let syntax = if let (_, Event::Start(Tag::CodeBlock(CodeBlockKind::Fenced(x)))) = &events[0] {
+        ss.find_syntax_by_token(x.split(',').next().unwrap())
+    } else {
+        None
+    };
+
+    let Some(syntax) = syntax else {
+        // If there is no language specifier, the whole codeblock should be translated.
```

Comment:
I just tested this and a quick look at the diff before/after this change shows something interesting:

````patch
+#: src/running-the-course/course-structure.md:57
+msgid ""
+"```shell\n"
+"cargo init concurrency\n"
+"cd concurrency\n"
+"cargo add tokio --features full\n"
+"cargo run\n"
+"```"
+msgstr ""
````

This seemed odd at first: why do we suddenly include a full code block?

The answer is in this line: before we used a simple heuristic and concluded that we could skip this block because it contains neither `"` nor `//`. Now, it is included verbatim because `shell` apparently isn't a recognized syntax name.

@dalance, could you restore the old heuristic in this case?

Reply:
I'll restore heuristic.

Reply:
Thanks a lot!